File: 2025/python/day09.py
from itertools import combinations

from utils import Point, get_line_data, get_nums, timeit


class Grid:
    def __init__(self, data: list[str]):
        self.points_set = {Point(get_nums(line)[0], get_nums(line)[1]) for line in data}
        self.points_ordered = [Point(get_nums(line)[0], get_nums(line)[1]) for line in data]
        self._vertical_lines = None
        self._horizontal_lines = None

    @staticmethod
    def _area_between_points(p1: Point, p2: Point) -> int:
        return (abs(p1.x - p2.x) + 1) * (abs(p1.y - p2.y) + 1)

    def _build_line_segments(self):
        if self._vertical_lines is not None:
            return

        self._vertical_lines = []
        self._horizontal_lines = []
        prev_point = self.points_ordered[-1]

        for point in self.points_ordered:
            if prev_point.x == point.x:
                # Vertical
                y1, y2 = prev_point.y, point.y
                self._vertical_lines.append((point.x, (y1, y2) if y1 < y2 else (y2, y1)))
            elif prev_point.y == point.y:
                # Horizontal
                x1, x2 = prev_point.x, point.x
                self._horizontal_lines.append(((x1, x2) if x1 < x2 else (x2, x1), point.y))
            prev_point = point

    def _check_lines_cross_rectangle(self, x_min: int, x_max: int, y_min: int, y_max: int) -> bool:
        # Check vertical
        for line_x, (line_y_start, line_y_end) in self._vertical_lines:
            if x_min < line_x < x_max:
                if line_y_start <= y_min < line_y_end or line_y_start < y_max <= line_y_end:
                    return False

        # Check horizontal
        for (line_x_start, line_x_end), line_y in self._horizontal_lines:
            if y_min < line_y < y_max:
                if line_x_start <= x_min < line_x_end or line_x_start < x_max <= line_x_end:
                    return False

        return True

    @timeit
    def part_one(self) -> int:
        return max(self._area_between_points(p1, p2) for p1, p2 in combinations(self.points_set, 2))

    @timeit
    def part_two(self) -> int:
        self._build_line_segments()

        n = len(self.points_ordered)
        areas_and_pairs = [
            (self._area_between_points(self.points_ordered[i], self.points_ordered[j]), i, j)
            for i in range(n)
            for j in range(i + 1, n)
        ]
        areas_and_pairs.sort(reverse=True, key=lambda x: x[0])

        for area, i, j in areas_and_pairs:
            p1, p2 = self.points_ordered[i], self.points_ordered[j]
            x_min, x_max = (p1.x, p2.x) if p1.x < p2.x else (p2.x, p1.x)
            y_min, y_max = (p1.y, p2.y) if p1.y < p2.y else (p2.y, p1.y)

            if self._check_lines_cross_rectangle(x_min, x_max, y_min, y_max):
                return area


@timeit
def main():
    data = get_line_data("09")
    grid = Grid(data)
    print(grid.part_one())
    print(grid.part_two())


if __name__ == '__main__':
    main()

File: 2025/rust/day09.rs
use crate::y2025::utils;
use itertools::Itertools;
use std::collections::HashSet;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
struct Point {
    x: i64,
    y: i64,
}

impl Point {
    fn new(x: i64, y: i64) -> Self {
        Point { x, y }
    }
}

struct Grid {
    points_set: HashSet<Point>,
    points_ordered: Vec<Point>,
    vertical_lines: Option<Vec<(i64, (i64, i64))>>,
    horizontal_lines: Option<Vec<((i64, i64), i64)>>,
}

impl Grid {
    fn new(data: &[String]) -> Self {
        let points_set: HashSet<Point> = data
            .iter()
            .map(|line| {
                let nums = utils::get_nums(line);
                Point::new(nums[0], nums[1])
            })
            .collect();

        let points_ordered: Vec<Point> = data
            .iter()
            .map(|line| {
                let nums = utils::get_nums(line);
                Point::new(nums[0], nums[1])
            })
            .collect();

        Grid {
            points_set,
            points_ordered,
            vertical_lines: None,
            horizontal_lines: None,
        }
    }

    fn area_between_points(p1: Point, p2: Point) -> i64 {
        ((p1.x - p2.x).abs() + 1) * ((p1.y - p2.y).abs() + 1)
    }

    fn build_line_segments(&mut self) {
        if self.vertical_lines.is_some() {
            return;
        }

        let mut vertical_lines = Vec::new();
        let mut horizontal_lines = Vec::new();
        let mut prev_point = self.points_ordered[self.points_ordered.len() - 1];

        for &point in &self.points_ordered {
            if prev_point.x == point.x {
                // Vertical line
                let y1 = prev_point.y;
                let y2 = point.y;
                let (y_start, y_end) = if y1 < y2 { (y1, y2) } else { (y2, y1) };
                vertical_lines.push((point.x, (y_start, y_end)));
            } else if prev_point.y == point.y {
                // Horizontal line
                let x1 = prev_point.x;
                let x2 = point.x;
                let (x_start, x_end) = if x1 < x2 { (x1, x2) } else { (x2, x1) };
                horizontal_lines.push(((x_start, x_end), point.y));
            }
            prev_point = point;
        }

        self.vertical_lines = Some(vertical_lines);
        self.horizontal_lines = Some(horizontal_lines);
    }

    fn check_lines_cross_rectangle(
        &self,
        x_min: i64,
        x_max: i64,
        y_min: i64,
        y_max: i64,
    ) -> bool {
        let vertical_lines = self.vertical_lines.as_ref().unwrap();
        let horizontal_lines = self.horizontal_lines.as_ref().unwrap();

        // Check vertical
        for &(line_x, (line_y_start, line_y_end)) in vertical_lines {
            if x_min < line_x && line_x < x_max {
                if (line_y_start <= y_min && y_min < line_y_end)
                    || (line_y_start < y_max && y_max <= line_y_end)
                {
                    return false;
                }
            }
        }

        // Check horizontal
        for &((line_x_start, line_x_end), line_y) in horizontal_lines {
            if y_min < line_y && line_y < y_max {
                if (line_x_start <= x_min && x_min < line_x_end)
                    || (line_x_start < x_max && x_max <= line_x_end)
                {
                    return false;
                }
            }
        }

        true
    }

    fn part_one(&self) -> i64 {
        self.points_set
            .iter()
            .combinations(2)
            .map(|v| Self::area_between_points(*v[0], *v[1]))
            .max()
            .unwrap_or(0)
    }

    fn part_two(&mut self) -> i64 {
        self.build_line_segments();

        let n = self.points_ordered.len();
        let mut areas_and_pairs: Vec<(i64, usize, usize)> = Vec::new();

        for i in 0..n {
            for j in (i + 1)..n {
                let area = Self::area_between_points(self.points_ordered[i], self.points_ordered[j]);
                areas_and_pairs.push((area, i, j));
            }
        }

        areas_and_pairs.sort_by(|a, b| b.0.cmp(&a.0));

        for (area, i, j) in areas_and_pairs {
            let p1 = self.points_ordered[i];
            let p2 = self.points_ordered[j];
            let (x_min, x_max) = if p1.x < p2.x { (p1.x, p2.x) } else { (p2.x, p1.x) };
            let (y_min, y_max) = if p1.y < p2.y { (p1.y, p2.y) } else { (p2.y, p1.y) };

            if self.check_lines_cross_rectangle(x_min, x_max, y_min, y_max) {
                return area;
            }
        }

        0
    }
}

pub fn main() {
    let data = utils::get_line_data("09");
    let mut grid = Grid::new(&data);
    println!("{}", grid.part_one());
    println!("{}", grid.part_two());
}

